File: backend/app/presentation/api/admin_users.py
from typing import Literal, Optional

from fastapi import APIRouter, Depends, HTTPException, Query

from app.application.interfaces.user import IUserRepository
from app.infrastructure.dependencies import get_user_repo
from app.infrastructure.security.authz import require_permissions
from app.presentation.schemas.admin_users import PagedUsersOut, SetRolesIn

router = APIRouter(prefix="/admin/users", tags=["AdminUsers"])


@router.put(
    "/{user_id}/roles",
    dependencies=[Depends(require_permissions(["admin_users:assign_role"]))],
)
def set_roles(user_id: int, payload: SetRolesIn, repo: IUserRepository = Depends(get_user_repo)):
    try:
        roles = repo.set_roles(user_id, payload.roles)
        return {"user_id": user_id, "roles": sorted(roles)}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get(
    "",
    response_model=PagedUsersOut,
    dependencies=[Depends(require_permissions(["admin_users:list"]))],
)
def list_users(
    q: Optional[str] = Query(default=None, min_length=1, max_length=200),
    role: Optional[str] = Query(default=None, max_length=64),
    sort_by: Literal["created_at", "email", "role"] = Query(default="created_at"),
    sort_dir: Literal["asc", "desc"] = Query(default="asc"),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=5, ge=1, le=100),
    repo: IUserRepository = Depends(get_user_repo),
):
    items, total = repo.list_admin_users(
        q=q,
        role=role,
        sort_by=sort_by,
        sort_dir=sort_dir,
        page=page,
        page_size=page_size,
    )
    return {"items": items, "total": total, "page": page, "page_size": page_size}

File: frontend/src/api/adminUsers.ts
import { api } from "./client";

export type AdminUserRow = {
  id: number;
  email: string;
  name?: string | null;
  roles: string[];
  created_at?: string | null;
  subscription_tier?: string | null;
};

export type ListUsersParams = {
  q?: string;
  role?: string;
  sort_by?: "created_at" | "email" | "role";
  sort_dir?: "asc" | "desc";
  page?: number;
  page_size?: number;
};

export type PagedUsers = {
  items: AdminUserRow[];
  total: number;
  page: number;
  page_size: number;
};

export async function listUsers(params?: ListUsersParams): Promise<PagedUsers> {
  const res = await api.get<PagedUsers>("/admin/users", { params });
  return res.data;
}

export async function setUserRoles(userId: number, roles: string[]) {
  const res = await api.put(`/admin/users/${userId}/roles`, { roles });
  return res.data;
}